src/k8s/cmd/k8sd/k8sd_cluster_recover.go

```diff
@@ -0,0 +1,409 @@
+					"Pre-recovery backup: %s, post-recovery backup: %s",
+					err,
+					k8sDqlitePreRecoveryTarball, k8sDqlitePostRecoveryTarball,
+				);
+			}
+			cmd.Printf("K8s-dqlite cluster changes applied; new database state saved to %s\n\n",
+					   k8sDqlitePostRecoveryTarball)
+			cmd.Printf("*Before* starting any cluster member, copy %s " +
+					   "on all remaining cluster members and extract the archive to %s.\n\n",
+					   k8sDqlitePostRecoveryTarball, clusterRecoverOpts.K8sDqliteStateDir)
+			cmd.Printf("Pre-recovery database backup: %s\n\n",
+					   k8sDqlitePreRecoveryTarball)
+
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVar(&clusterRecoverOpts.K8sDqliteStateDir, "state-dir",
+						  "", "k8s-dqlite datastore location")
+
+	return cmd
+}
+
+func removeYamlHelperComments(content []byte) []byte {
+	pattern := fmt.Sprintf("(?s).*?\n%s *\n", yamlHelperCommentFooter)
+	re := regexp.MustCompile(pattern)
+	return re.ReplaceAll(content, nil)
+}
+
+func removeYamlHelperCommentsFromFile(path string) error {
+	content, err := os.ReadFile(path)
+	if err != nil {
+		return fmt.Errorf("Could not retrieve file contents: %s. Error: %w.", path, err);
+	}
+
+	content = removeYamlHelperComments(content)
+
+	return os.WriteFile(path, content, os.FileMode(0o644))
+}
+
+func recoveryCmdPrechecks() error {
+	if !termios.IsTerminal(unix.Stdin) {
+		return fmt.Errorf("This command is meant to be run in an interactive terminal.")
+	}
+
+	reader := bufio.NewReader(os.Stdin)
+	fmt.Print(recoveryConfirmation)
+
+	input, err := reader.ReadString('\n')
+	if err != nil {
+		return fmt.Errorf("Couldn't read user input: %w.", err)
+	}
+	input = strings.TrimSuffix(input, "\n")
+
+	if strings.ToLower(input) != "yes" {
+		return fmt.Errorf("Cluster edit aborted; no changes made.")
+	}
+
+	return nil
+}
+
+func ensureK8sDqliteMembersStopped(members []cluster.DqliteMember) error {
+	crt := path.Join(clusterRecoverOpts.K8sDqliteStateDir, "cluster.crt")
+	key := path.Join(clusterRecoverOpts.K8sDqliteStateDir, "cluster.key")
+
+	keypair, err := tls.LoadX509KeyPair(crt, key)
+	if err != nil {
+		return fmt.Errorf("Could not load k8s-dqlite certificates. Error: %w.", err)
+	}
+
+	data, err := os.ReadFile(crt)
+	if err != nil {
+		return fmt.Errorf("Could not read k8s-dqlite certificate. Error: %w.", err)
+	}
+
+	pool := x509.NewCertPool()
+	if !pool.AppendCertsFromPEM(data) {
+		return fmt.Errorf("Invalid k8s-dqlite certificate.")
+	}
+
+	dial := client.DialFuncWithTLS(client.DefaultDialFunc, app.SimpleDialTLSConfig(keypair, pool))
+
+	// We'll dial all members concurrently, passing the reachable addresses
+	// through a channel.
+	availableMembers := []string{}
+	c := make(chan string)
+	for _, member := range members {
+		go func(dialFunc client.DialFunc, addr string) {
+			// We expect the member nodes to be unavailable, let's not wait
+			// more than 3 seconds during this sanity test.
+			ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
```

Comment:
this should be the `cmd.Context` otherwise cancelling the command would not stop this santity check.